Services/communication_service.c

```diff
@@ -37,155 +38,205 @@ SAT_returnState communication_service_app(csp_packet_t *packet) {
-      set_packet_length(packet, sizeof(int8_t) + 2*sizeof(uint8_t) + 1);
+      memcpy(&packet->data[OUT_DATA_BYTE], &S_status, sizeof(S_status));
+      set_packet_length(packet, sizeof(int8_t) + sizeof(S_status) + 1);
 
       if (queue_response(packet) != SATR_OK) {
         return SATR_ERROR;
       }
       break;
 
     case GET_TR:
-        HAL_S_getTR(&S_buffer.transmit);
-        S_buffer.transmit = csp_hton32(S_buffer.transmit);
+        status = HAL_S_getTR(&S_TR);
+        if (sizeof(S_TR) + 1 > csp_buffer_data_size()) {
+          return SATR_ERROR;
+        }
+        S_TR.transmit = csp_hton32((uint32_t)S_TR.transmit);
         // what to do for int?
-        status = 0;
         memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
-        memcpy(&packet->data[OUT_DATA_BYTE], &S_buffer.transmit, sizeof(int));
-        set_packet_length(packet, sizeof(int8_t) + sizeof(int) + 1);
+        memcpy(&packet->data[OUT_DATA_BYTE], &S_TR, sizeof(S_TR));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(S_TR) + 1);
         if (queue_response(packet) != SATR_OK) {
           return SATR_ERROR;
         }
         break;
 
+    case GET_HK:
+        status = HAL_S_getHK (&S_HK);
+        if (sizeof(S_HK) + 1 > csp_buffer_data_size()) {
+          return SATR_ERROR;
+        }
+        /*Use va_list from stdarg for various number of arguments. */
+        S_HK.Output_Power = csp_htonflt(S_HK.Output_Power);
+        S_HK.PA_Temp = csp_htonflt(S_HK.PA_Temp);
+        S_HK.Top_Temp = csp_htonflt(S_HK.Top_Temp);
+        S_HK.Bottom_Temp = csp_htonflt(S_HK.Bottom_Temp);
+        S_HK.Bat_Current = csp_htonflt(S_HK.Bat_Current);
+        S_HK.Bat_Voltage = csp_htonflt(S_HK.Bat_Voltage);
+        S_HK.PA_Current = csp_htonflt(S_HK.PA_Current);
+        S_HK.PA_Voltage = csp_htonflt(S_HK.PA_Voltage);
+
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        memcpy(&packet->data[OUT_DATA_BYTE], &S_HK, sizeof(S_HK));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(S_HK) + 1);
+
+        if (queue_response(packet) != SATR_OK) {
+          return SATR_ERROR;
+        }
+              break;
+
+    case GET_BUFFER:
+      SID = packet->data[SID_byte];
+      status = HAL_S_getBuffer(SID, &S_buffer);
+      if (sizeof(S_buffer) + 1 > csp_buffer_data_size()) {
+        return SATR_ERROR;
+      }
+      S_buffer.pointer[SID] = csp_hton16(S_buffer.pointer[SID]);
+      memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE], &S_buffer.pointer[SID], sizeof(S_buffer.pointer[SID]));
+      set_packet_length(packet, sizeof(int8_t) + sizeof(S_buffer.pointer[SID]) + 1);
+      if (queue_response(packet) != SATR_OK) {
+        return SATR_ERROR;
+      }
+      break;
 
-/*    case SET_FREQ:
-      cnv8_32(&packet->data[IN_DATA_BYTE],&S_config.S_freq); //check what happens if skipped
-      S_config.S_freq = csp_ntoh32(S_config.S_freq);
```

Comment:
Since it'll be my first time using a mutex, I might ask you a question about using them in the set functions, soon. I found some resources and examples to check out first.